okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java

```diff
@@ -144,53 +181,116 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
   }
 
   /**
-   * Parse RFC 7617 challenges, also wrong ordered ones.
-   * This API is only interested in the scheme name and realm.
+   * Parse RFC 7235 challenges.
+   * If a challenge uses the {@code token68} variant instead of auth params,
+   * there is exactly one auth param in the challenge at key {@code null}.
+   * Invalid headers and challenges are ignored.
+   * No semantic validation is done, for example that {@code Basic} auth must have a
+   * {@code realm} auth param, this is up to the caller that interprets these challenges.
    */
   public static List<Challenge> parseChallenges(Headers responseHeaders, String challengeHeader) {
-    // auth-scheme = token
-    // auth-param  = token "=" ( token | quoted-string )
-    // challenge   = auth-scheme 1*SP 1#auth-param
-    // realm       = "realm" "=" realm-value
-    // realm-value = quoted-string
     List<Challenge> challenges = new ArrayList<>();
     List<String> authenticationHeaders = responseHeaders.values(challengeHeader);
+headerLoop:
     for (String header : authenticationHeaders) {
-      int index = header.indexOf(' ');
-      if (index == -1) continue;
-
-      String scheme = header.substring(0, index);
-      String realm = null;
-      String charset = null;
-
-      Matcher matcher = PARAMETER.matcher(header);
-      for (int i = index; matcher.find(i); i = matcher.end()) {
-        if (header.regionMatches(true, matcher.start(1), "realm", 0, 5)) {
-          realm = matcher.group(3);
-        } else if (header.regionMatches(true, matcher.start(1), "charset", 0, 7)) {
-          charset = matcher.group(3);
-        }
+      List<Challenge> headerChallenges = CHALLENGES_CACHE.get(header);
```

Comment:
The cache makes sense given the repeated nature of the non-preemptive challenges.  But I don't think we should have it in an unbounded form.

I think we should either restrict it in some way e.g. TTL, size.  Or just accept the wasted effort, which is hopefully noise compared to the network cost.

Reply:
Yeah, I don’t want a ConcurrentHashMap here. The solution is to make decoding challenges fast & simple, not to cache it!

Reply:
So you want me to simply remove the cache?

Reply:
Ok, removed